### PR TITLE
Improve error message for import expression

### PR DESCRIPTION
**Pull request by andrey-zherikov**
Better error message if import expression has absolute path or a reference to parent directory.

### CODE REVIEW

File: test/fail_compilation/fail21227.d [hunk 1]
```diff
@@ -1,20 +0,0 @@
...
----
-fail_compilation\fail21227.d(2): Error: file `"/abs/path/to/file.txt"` cannot be found or not in a path specified with `-J`
-fail_compilation\fail21227.d(2):        Path(s) searched (as provided by `-J`):
-fail_compilation\fail21227.d(2):        [0]: `fail_compilation`
-fail_compilation\fail21227.d(3): Error: file `"../file.txt"` cannot be found or not in a path specified with `-J`
-fail_compilation\fail21227.d(3):        Path(s) searched (as provided by `-J`):
-fail_compilation\fail21227.d(3):        [0]: `fail_compilation`
-fail_compilation\fail21227.d(4): Error: file `"path/to/parent/../file.txt"` cannot be found or not in a path specified with `-J`
-fail_compilation\fail21227.d(4):        Path(s) searched (as provided by `-J`):
-fail_compilation\fail21227.d(4):        [0]: `fail_compilation`
```
[the remark this answers is not in the document]
Comment (andrey-zherikov) [reply test/fail_compilation/fail21227.d:13]: Added

**Comment by Geod24** [comment]
While any effort to improve error message is welcome, I am not so convinced by the diff in this PR.

Here's a few points that IMO needs improvement:
- The Windows error message is way too generic, mentioning all possible disallowed characters, and mixed with another class of errors (using `..`). It's also too long, as a result. The user doesn't really know what went wrong. A better error message would say exactly what went wrong and why, for example:
```
Error: `"C:\foo\bar\hello<world.txt"` is not a valid filename on this platform
           Character `'<'` is reserved by the Win32 API and cannot be used
```
This gives the user a much clearer indication of what went wrong and what needs to be fixed. It also create a smaller error message by using `errorSupplemental`.

- There is nothing inherently wrong with using `..` or even an absolute path. What if someone writes a tool that parses C headers at compile time to generate D declarations ? You would want to be able to `import("/usr/include/stdio.h")` as long as the user gave consent (by using `-J/usr/inlcude/`). In this regard, I don't think erroring out on absolute path or the presence of `..` is wise.

**Comment by andrey-zherikov** [comment]
> * The Windows error message is way too generic, mentioning all possible disallowed characters, and mixed with another class of errors (using `..`). It's also too long, as a result. The user doesn't really know what went wrong. A better error message would say exactly what went wrong and why, for example:
> 
> ```
> Error: `"C:\foo\bar\hello<world.txt"` is not a valid filename on this platform
>            Character `'<'` is reserved by the Win32 API and cannot be used
> ```
> 
> This gives the user a much clearer indication of what went wrong and what needs to be fixed. It also create a smaller error message by using `errorSupplemental`.

Thanks for suggestion, I'll do that.

> * There is nothing inherently wrong with using `..` or even an absolute path. What if someone writes a tool that parses C headers at compile time to generate D declarations ? You would want to be able to `import("/usr/include/stdio.h")` as long as the user gave consent (by using `-J/usr/inlcude/`). In this regard, I don't think erroring out on absolute path or the presence of `..` is wise.

The goal of this PR is to improve error message for these cases but not to change how dmd is looking for a file (I believe such change should be done in separate PR).

Providing absolute path in import expression might be legitimate and solve the issue where the same file is located in two or more `-J` directories. But IMO this is not really useful because the author of the code (who writes import expression) should agree with the one who compiles the code where imported files are located. If it's one person and software is being compiled locally then it's fine but consider the case where CI tool is used to build - you can't guarantee any path to your workspace there.

Regarding presence of `..` in a path: I agree that there is nothing wrong with using `..` in a path unless it points outside of `-J` path (i.e. normalized path starts with `..`), but again it should be in separate PR.

**Comment by andrey-zherikov** [comment]
> * The Windows error message is way too generic, mentioning all possible disallowed characters, and mixed with another class of errors (using `..`). It's also too long, as a result. The user doesn't really know what went wrong. A better error message would say exactly what went wrong and why, for example:
> 
> ```
> Error: `"C:\foo\bar\hello<world.txt"` is not a valid filename on this platform
>            Character `'<'` is reserved by the Win32 API and cannot be used
> ```
> 
> This gives the user a much clearer indication of what went wrong and what needs to be fixed. It also create a smaller error message by using `errorSupplemental`.


Done.

**Comment by andrey-zherikov** [comment]
I removed `FileName.safeSearchPath` and `FileName.safePath` as they are used in one place only and path check they are doing is already performed in `expressionsem.d` (place where they are called).

**Comment by andrey-zherikov** [comment]
Is there anything required from me so this PR can be landed?

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/frontend.h [hunk 2]
```diff
@@ -1173,7 +1173,6 @@ struct FileName
     static bool equalsExt(const char* name, const char* ext);
     bool equalsExt(const char* ext) const;
     static const char* searchPath(Array<const char* >* path, const char* name, bool cwd);
-    static const char* safeSearchPath(Array<const char* >* path, const char* name);
```
Comment (MoonlightSentinel) [inline src/dmd/frontend.h:1176]: This change was not applied to `filename.h`.

Fix #11932

**Pull request merged**